File: Solutions/exo3/src/task2/knapsack.js
function knapsack01(weights, values, capacity) {
  const n = weights.length;
  const dp = Array(n + 1)
    .fill(null)
    .map(() => Array(capacity + 1).fill(0));

  for (let i = 1; i <= n; i++) {
    for (let w = 0; w <= capacity; w++) {
      dp[i][w] = dp[i - 1][w];

      if (weights[i - 1] <= w) {
        dp[i][w] = Math.max(
          dp[i][w],
          dp[i - 1][w - weights[i - 1]] + values[i - 1]
        );
      }
    }
  }

  const selected = [];
  let w = capacity;
  for (let i = n; i > 0; i--) {
    if (dp[i][w] !== dp[i - 1][w]) {
      selected.push(i - 1);
      w -= weights[i - 1];
    }
  }

  selected.reverse();
  return { maxValue: dp[n][capacity], selectedItems: selected };
}

function knapsack01Optimized(weights, values, capacity) {
  const n = weights.length;
  const dp = Array(capacity + 1).fill(0);

  for (let i = 0; i < n; i++) {
    for (let w = capacity; w >= weights[i]; w--) {
      dp[w] = Math.max(dp[w], dp[w - weights[i]] + values[i]);
    }
  }

  return dp[capacity];
}

const weights = [2, 3, 4, 5];
const values = [3, 4, 5, 6];
const capacity = 8;

const result = knapsack01(weights, values, capacity);
console.log("Maximum value:", result.maxValue);
console.log("Selected items (indices):", result.selectedItems);
console.log(
  "Selected weights:",
  result.selectedItems.map((i) => weights[i])
);
console.log(
  "Selected values:",
  result.selectedItems.map((i) => values[i])
);
console.log(
  "Total weight:",
  result.selectedItems.reduce((sum, i) => sum + weights[i], 0)
);

console.log(
  "\nOptimized space solution max value:",
  knapsack01Optimized(weights, values, capacity)
);

File: 25_2/s_11/exo3/knapsack.py
def knapsack_01(weights, values, capacity):
    """
    Solve 0/1 knapsack problem using dynamic programming
    Returns: max_value, selected_items
    """
    n = len(weights)
    # DP table initialization
    dp = [[0] * (capacity + 1) for _ in range(n + 1)]
    
    # Fill DP table
    for i in range(1, n + 1):
        for w in range(1, capacity + 1):
            if weights[i-1] <= w:
                # Maximum of: take item or skip item
                dp[i][w] = max(dp[i-1][w], 
                              values[i-1] + dp[i-1][w - weights[i-1]])
            else:
                dp[i][w] = dp[i-1][w]
    
    # Backtrack to find selected items
    selected_items = []
    w = capacity
    for i in range(n, 0, -1):
        if dp[i][w] != dp[i-1][w]:
            selected_items.append(i-1)
            w -= weights[i-1]
    
    return dp[n][capacity], selected_items

def knapsack_optimized(weights, values, capacity):
    """
    Space-optimized version using O(capacity) memory
    """
    n = len(weights)
    dp = [0] * (capacity + 1)
    
    for i in range(n):
        # Iterate backwards to avoid overwriting
        for w in range(capacity, weights[i] - 1, -1):
            dp[w] = max(dp[w], values[i] + dp[w - weights[i]])
    
    return dp[capacity]

# Example usage
if __name__ == "__main__":
    weights = [2, 3, 4, 5]
    values = [3, 4, 5, 6]
    capacity = 5
    
    max_value, items = knapsack_01(weights, values, capacity)
    max_value_opt = knapsack_optimized(weights, values, capacity)
    
    print("0/1 Knapsack Problem Solution")
    print(f"Weights: {weights}")
    print(f"Values: {values}")
    print(f"Capacity: {capacity}")
    print(f"Maximum value: {max_value}")
    print(f"Selected items: {items}")
    print(f"Optimized version result: {max_value_opt}")